lib/Parse/ParseDecl.cpp

```diff
@@ -2005,13 +2005,28 @@ ParserStatus Parser::parseDecl(SmallVectorImpl<Decl*> &Entries,
                                                   false);
         }
       }
+      if (CurDeclContext) {
+        auto ty =
+          CurDeclContext->getAsNominalTypeOrNominalTypeExtensionContext();
+        if (ty) {
+          diagnose(ty->getLoc(), diag::note_in_decl_extension, false,
+                   ty->getNameStr());
+        } else if (auto ed = dyn_cast<ExtensionDecl>(CurDeclContext)) {
+          if (auto extension = ed->getExtendedType()) {
```

Comment:
This whole branch is dead because getAsNominalTypeOrNominalTypeExtensionContext() returns the extended type if we have an extension -- you should just move the ```isa<ExtensionDecl>()``` inside the ```if (ty) {```.